File: app/src/main/java/com/estasvegano/android/estasvegano/entity/Category.java
package com.estasvegano.android.estasvegano.entity;

import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;

import java.util.List;

import auto.parcel.AutoParcel;

@AutoParcel
@JsonDeserialize(builder = AutoParcel_Category.Builder.class)
public abstract class Category implements Parcelable {

    @NonNull
    public static Builder builder() {
        return new AutoParcel_Category.Builder();
    }

    @NonNull
    public abstract Builder toBuilder();

    @JsonProperty("id")
    @NonNull
    public abstract long id();

    @JsonProperty("title")
    @NonNull
    public abstract String title();

    @JsonProperty("children")
    @Nullable
    public abstract List subCategories();

    @JsonProperty("parent")
    @Nullable
    public abstract Category parent();

    public boolean isLowLevel() {
        return subCategories() == null || subCategories().size() == 0;
    }

    @AutoParcel.Builder
    public static abstract class Builder {

        @JsonProperty("id")
        @NonNull
        public abstract Builder id(long id);

        @JsonProperty("title")
        @NonNull
        public abstract Builder title(@NonNull String title);

        @JsonProperty("children")
        @NonNull
        public abstract Builder subCategories(@Nullable List subCategories);

        @JsonProperty("parent")
        @NonNull
        public abstract Builder parent(@Nullable Category parent);

        @NonNull
        public abstract Category build();
    }
}

File: app/src/main/java/com/estasvegano/android/estasvegano/entity/Category.kt
package com.estasvegano.android.estasvegano.entity

import android.os.Parcel
import android.os.Parcelable

data class Category(
        val id: Long,
        val title: String,
        val subCategories: List<Category>?,
        val parent: Category?
) : Parcelable {

    val isLowLevel: Boolean
        get() = subCategories?.isNotEmpty() ?: true

    companion object {
        @JvmField val CREATOR: Parcelable.Creator<Category> = object : Parcelable.Creator<Category> {
            override fun createFromParcel(source: Parcel): Category = Category(source)
            override fun newArray(size: Int): Array<Category?> = arrayOfNulls(size)
        }
    }

    constructor(source: Parcel) : this(
            source.readLong(),
            source.readString(),
            source.createTypedArrayList(Category.CREATOR),
            source.readParcelable<Category>(Category::class.java.classLoader)
    )

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeLong(id)
        dest.writeString(title)
        dest.writeTypedList(subCategories)
        dest.writeParcelable(parent, 0)
    }
}
